src/org/traccar/protocol/LaipacProtocolDecoder.java

```diff
@@ -88,28 +101,86 @@ protected Object decode(
         dateBuilder.setDateReverse(parser.nextInt(0), parser.nextInt(0), parser.nextInt(0));
         position.setTime(dateBuilder.getDate());
 
-        String type = parser.next();
+        String eventCode = parser.next();
+        String decodedAlarm = decodeAlarm(eventCode);
+        if (decodedAlarm != null) {
+            position.set(Position.KEY_ALARM, decodeAlarm(eventCode));
+        }
+        position.set(Position.KEY_EVENT, eventCode);
+
+        double batteryVoltage = parser.nextDouble();
+        if (parser.next().isEmpty()) {
+            parser.next();
+            batteryVoltage *= 0.001;
+        } else {
+            batteryVoltage += parser.nextDouble() * 0.001;
+        }
+        position.set(Position.KEY_BATTERY, batteryVoltage);
+
+        position.set(Position.KEY_TOTAL_DISTANCE, parser.nextDouble());
+        position.set(Position.KEY_GPS, parser.nextInt());
+        position.set(Position.PREFIX_ADC + 1, parser.nextDouble() * 0.001);
+        position.set(Position.PREFIX_ADC + 2, parser.nextDouble() * 0.001);
+
+        String cellNetCodeString = parser.next();
+        String cellIdString = parser.next();
+        String countryCodeString = parser.next();
+        String operatorCodeString = parser.next();
+        if (cellNetCodeString != null
+                && cellIdString != null
+                && countryCodeString != null
+                && operatorCodeString != null) {
+            int cellNetCode = Integer.parseInt(cellNetCodeString, 16);
+            Long cellId = Long.parseLong(cellIdString, 16);
+            int countryCode = Integer.parseInt(countryCodeString);
+            int operatorCode = Integer.parseInt(operatorCodeString);
+            position.setNetwork(new Network(CellTower.from(countryCode, operatorCode, cellNetCode, cellId)));
+        }
+
         String checksum = parser.next();
+        String result = sentence.replaceAll("^\\$(.*)\\*[0-9a-fA-F]{2}$", "$1");
```

Comment:
What is this for? I suggest we remove checksum checking as we don't use it in other places. TCP has own checksum which is more than enough.

Reply:
@tananaev I removed it but me and my engineering team are not sure why you wouldn't check it. Is there any special reason why you don't check this?

Reply:
It's about consistency and performance.

If you want to verify checksum, we should add it to all other protocol, not just this one.

Second is performance. It just waste of CPU time. I don't see any value in checksum because TCP and UDP protocol already have a better checksum. For example, you don't see checksum in HTTP protocol because it's an application level protocol and integrity of the data should be checked on the transport layer (which is TCP).